python/ray/tests/test_memory_deadlock.py

```diff
@@ -0,0 +1,259 @@
+        self.mem = []
+        self.barrier = barrier
+
+    def perform_allocations(self, actor_bytes, nested_task_bytes):
+        self.mem = alloc_mem(actor_bytes)
+        if self.barrier:
+            ray.get(self.barrier.wait_all_done.remote())
+        ray.get(allocate_memory.options(max_retries=0).remote(nested_task_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_actor_with_nested_task(
+    ray_with_memory_monitor,
+):
+    leaker = ActorWithNestedTask.options(max_restarts=0, max_task_retries=0).remote()
+    actor_bytes = get_additional_bytes_to_reach_memory_usage_pct(
+        (memory_usage_threshold_fraction + 0.25) / 2
+    )
+    nested_task_bytes = (
+        get_additional_bytes_to_reach_memory_usage_pct(
+            memory_usage_threshold_fraction + 0.25
+        )
+        - actor_bytes
+    )
+    with pytest.raises(ray.exceptions.RayTaskError) as _:
+        ray.get(leaker.perform_allocations.remote(actor_bytes, nested_task_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_two_sets_of_actor_with_nested_task(
+    ray_with_memory_monitor,
+):
+    barrier = BarrierActor.options(
+        max_restarts=0, max_task_retries=0, max_concurrency=2
+    ).remote(2)
+    leaker1 = ActorWithNestedTask.options(max_restarts=0, max_task_retries=0).remote(
+        barrier
+    )
+    leaker2 = ActorWithNestedTask.options(max_restarts=0, max_task_retries=0).remote(
+        barrier
+    )
+    bytes_first = get_additional_bytes_to_reach_memory_usage_pct(0.25)
+    quarter_mem = get_additional_bytes_to_reach_memory_usage_pct(0.5) - bytes_first
+    ref1 = leaker1.perform_allocations.remote(bytes_first, quarter_mem)
+    ref2 = leaker2.perform_allocations.remote(quarter_mem, 0)
+
+    num_failed = 0
+    try:
+        ray.get(ref1)
+    except ray.exceptions.RayTaskError:
+        num_failed += 1
+    try:
+        ray.get(ref2)
+    except ray.exceptions.RayTaskError:
+        num_failed += 1
+    assert num_failed == 1
+
+
+@ray.remote
+def task_with_nested_task(task_bytes, nested_task_bytes, barrier=None):
+    dummy = alloc_mem(task_bytes)
+    if barrier:
+        ray.get(barrier.wait_all_done.remote())
+    ray.get(allocate_memory.options(max_retries=0).remote(nested_task_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_multiple_tasks_with_nested_task(
+    ray_with_memory_monitor,
+):
+    """task_with_nested_task allocates a block of memor, then runs
+    a nested task which also allocates a block memory.
+    This test runs two instances of task_with_nested_task.
+    We expect the second one to fail."""
+    bytes_first = get_additional_bytes_to_reach_memory_usage_pct(0.3)
```

Comment:
why not do the following

allocate the actors each with (threshold - 0.1) / 2

allocate the nested tasks with mem(threshold) - mem(threshold - 0.2)